posts/map_blocks_example.md

```diff
@@ -0,0 +1,191 @@
+---
+author: Matt Long
+date: 2021-4-22
+tags: xarray, dask, pop
+---
+
+# How to Use `xarray.map_blocks` for Vertical Interpolation of a 3D Field
+
+Within this example, we cover how to use `xarray.map_blocks` to calculate the mixed-layer depth within the CESM POP model output.
+
+This calculation is "embarassingly parallel" such that each calculation is done within a single a column. The calculation **_should_** be easily computed within each column across the model domain. This is where `map_blocks` can be used to improve the performance of this metric.
+
+## Imports and Data Ingestion
+
+We use the following libraries/packages within this example
+
+```python
+import os
+
+from scipy import interpolate
+import numpy as np
+import xarray as xr
+
+import matplotlib.pyplot as plt
+
+import pop_tools
+```
+
+Next, we read in the data using `xarray` and subset for temperature and salinity, subsetting for a smaller portion of the domain for testing purposes
+
+```
+ds = xr.merge((
+    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/TEMP-presentday-monclim.nc'),
+    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/SALT-presentday-monclim.nc'),
+))
+
+
+ds = ds[['TEMP', 'SALT']]
+
+# subset for testing purposes
+ds = ds.isel(member_id=slice(0, 2), nlat=slice(2, 5), nlon=slice(1, 5))
+```
+
+## Data Operation
+
+Next, we setup a function to compute the mixed layer depth (MLD). The MLD is defined as the point in the water column where the potential density (`sigma`) exceeds the surface density by a specified threshold (`dsigma`).
+
+```python
+def mld_dsigma(SALT, TEMP, dsigma=0.03, rho_chunks={'nlat': 16, 'nlon': 16}):
+    """
+    Compute MLD based on ∆σ criterion. Uses xarray.map_blocks.
+
+    Parameters
+    ----------
+
+    SALT : xarray.DataArray
+      Salinity
+    TEMP : xarray.DataArray
+      Potential temperature
+    dsigma : float, optional
+      The value for ∆σ.
+    rho_chunks: dictionary, optional
+      Dimension chunk parameters to chunk rho by
+
+    Returns
+    -------
+
+    MLD : xarray.DataArray
+      The MLD (m) defined as the point in the water column where
+      density exceeds rho[0] + dsigma.
+    """
+
+    # determine dimensionality
+    dims_in = SALT.dims
+    assert dims_in == TEMP.dims, 'dimension mismatch'
+    assert 'z_t' in SALT.coords, 'z_t not found in SALT coords'
+
+    # drop ancillary coordinates (this may not be necessary)
+    SALT = SALT.reset_coords(drop=True)
```

Comment:
Absolutely necessary. `map_blocks` takes a block of each dask array, wraps it up in xarray metadata, then wraps all that in to a dataset and sends it to your function. You should only send the data you need to avoid a lot of unnecessary computations.